Test-Scenario-1:
  id: ts_01
  path: tests/hydratk/lib/network/ftp/test_sftp_client.jedi
  name: test_sftp_client
  desc: Unit tests for module hydratk/lib/network/ftp/sftp_client.py
  author: Petr Rasek <bowman@hydratk.org>
  version: 1.0
  
  pre-req: |
    import yodalib.hydratk.lib.network.ftp.sftp_client as ftp    
    import os
    
  Test-Case-1:
    id: tc_01
    name: connect
    desc: Tests for method connect
  
    Test-Condition-1: 
      id: tco_01
      name: min_param
      desc: Connect with minimum parameters
      test: |
        c = ftp.TestFTPClient()
        host = 'localhost'
        c._events = []
        res = c.connect(host)
                         
      validate: |  
        expected = [True, host, 22, '/']
        current = [res, c._client.host, c._client.port, c._client.path]
        this.test_result = (current == expected)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(expected, current)  
        events = ['ftp_before_connect', 'ftp_before_change_dir', 'ftp_after_connect']
        this.test_result = (c._events == events)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(events, c._events)    
        
    Test-Condition-2: 
      id: tco_02
      name: max_param
      desc: Connect with maximum parameters
      test: |
        c = ftp.TestFTPClient()
        [host, port, user, passw, path] = ['localhost', 222, None, 'password', '/home/hydra']
        res = c.connect(host, port, user, passw, path)
                         
      validate: |  
        expected = [True, host, port, user, passw, path]
        current = [res, c._client.host, c._client.port, c._client.user, c._client.passw, c._client.path]
        this.test_result = (current == expected)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(expected, current)            
        
    Test-Condition-3: 
      id: tco_03
      name: exception
      desc: Connect exception
      test: |
        res = c.connect(host, port, user, passw, ex=True)
                         
      validate: |  
        this.test_result = (res == False)
        assert (this.test_result == True), 'Expected exception' 
        
  Test-Case-2:
    id: tc_02
    name: disconnect
    desc: Tests for method disconnect
  
    Test-Condition-1: 
      id: tco_01
      name: disconnect
      desc: Disconnect
      test: |
        c = ftp.TestFTPClient()
        res = c.disconnect()
                         
      validate: |  
        expected = True
        current = res
        this.test_result = (current == expected)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(expected, current)   
        
    Test-Condition-2: 
      id: tco_02
      name: exception
      desc: Disconnect exception
      test: |
        res = c.disconnect(ex=True)
                         
      validate: |  
        this.test_result = (res == False)
        assert (this.test_result == True), 'Expected exception'           
        
  Test-Case-3:
    id: tc_03
    name: list_dir
    desc: Tests for method list_dir
  
    Test-Condition-1: 
      id: tco_01
      name: list_dir
      desc: List directory
      test: |
        c = ftp.TestFTPClient()
        names = c.list_dir()
                         
      validate: |  
        expected = 4
        current = len(names)
        this.test_result = (current == expected)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(expected, current)             
        
    Test-Condition-2: 
      id: tco_02
      name: exception
      desc: List directory exception
      test: |
        names = c.list_dir(ex=True)
                         
      validate: |  
        this.test_result = (names == None)
        assert (this.test_result == True), 'Expected exception'    
        
  Test-Case-4:
    id: tc_04
    name: change_dir
    desc: Tests for method change_dir
  
    Test-Condition-1: 
      id: tco_01
      name: change_dir
      desc: Change directory
      test: |
        c = ftp.TestFTPClient()
        c._events = []
        path = '/home/hydratk'
        res = c.change_dir(path)
                         
      validate: |  
        expected = [True, path]
        current = [res, c._client.path]
        this.test_result = (current == expected)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(expected, current) 
        events = ['ftp_before_change_dir']
        this.test_result = (c._events == events)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(events, c._events)                  
        
    Test-Condition-2: 
      id: tco_02
      name: exception
      desc: Change directory exception
      test: |
        res = c.change_dir(path, ex=True)
                         
      validate: |  
        this.test_result = (res == False)
        assert (this.test_result == True), 'Expected exception'    
        
  Test-Case-5:
    id: tc_05
    name: download_file
    desc: Tests for method download_file
  
    Test-Condition-1: 
      id: tco_01
      name: min_params
      desc: Download file with minimum params
      test: |
        c = ftp.TestFTPClient()
        c._events = []
        remote_path = '/home/hydratk/tmp.txt'
        res = c.download_file(remote_path)
                         
      validate: |  
        expected = True
        current = res
        this.test_result = (current == expected)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(expected, current) 
        events = ['ftp_before_download_file', 'ftp_after_download_file']
        this.test_result = (c._events == events)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(events, c._events)        
        
    Test-Condition-2: 
      id: tco_02
      name: max_params
      desc: Download file with maximum params
      test: |
        c = ftp.TestFTPClient()
        local_path = '..'
        res = c.download_file(remote_path, local_path)
                         
      validate: |  
        expected = True
        current = res
        this.test_result = (current == expected)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(expected, current)       
        
    Test-Condition-3: 
      id: tco_03
      name: exception
      desc: Download file exception
      test: |
        res = c.download_file(remote_path, ex=True)
                         
      validate: |  
        this.test_result = (res == False)
        assert (this.test_result == True), 'Expected exception'     
        
  Test-Case-6:
    id: tc_06
    name: upload_file
    desc: Tests for method upload_file
  
    Test-Condition-1: 
      id: tco_01
      name: min_params
      desc: Upload file with minimum params
      test: |
        c = ftp.TestFTPClient()
        c._events = []
        local_path = 'tmp.txt'
        with open(local_path, 'w') as f:
            f.write('hydra')
        res = c.upload_file(local_path)        
                         
      validate: |  
        expected = True
        current = res
        this.test_result = (current == expected)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(expected, current) 
        events = ['ftp_before_upload_file', 'ftp_after_upload_file']
        this.test_result = (c._events == events)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(events, c._events)           
        os.remove(local_path)
        
    Test-Condition-2: 
      id: tco_02
      name: max_params
      desc: Upload file with maximum params
      test: |
        c = ftp.TestFTPClient()
        remote_path = '/home/hydratk'
        with open(local_path, 'w') as f:
            f.write('hydra')
        res = c.upload_file(local_path, remote_path)               
                         
      validate: |  
        expected = True
        current = res
        this.test_result = (current == expected)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(expected, current)        
        os.remove(local_path)  
        
    Test-Condition-3: 
      id: tco_03
      name: exception
      desc: Upload file exception
      test: |
        with open(local_path, 'w') as f:
            f.write('hydra')      
        res = c.upload_file(local_path, ex=True)
                         
      validate: |  
        this.test_result = (res == False)
        assert (this.test_result == True), 'Expected exception'   
        os.remove(local_path)          
        
  Test-Case-7:
    id: tc_07
    name: delete_file
    desc: Tests for method delete_file
  
    Test-Condition-1: 
      id: tco_01
      name: delete_file
      desc: Delete file
      test: |
        c = ftp.TestFTPClient()
        c._events = []
        path = '/home/hydratk/tmp.txt'
        res = c.delete_file(path)
                         
      validate: |  
        expected = True
        current = res
        this.test_result = (current == expected)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(expected, current) 
        events = ['ftp_before_delete_file']
        this.test_result = (c._events == events)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(events, c._events)                  
        
    Test-Condition-2: 
      id: tco_02
      name: exception
      desc: Delete file exception
      test: |
        res = c.delete_file(path, ex=True)
                         
      validate: |  
        this.test_result = (res == False)
        assert (this.test_result == True), 'Expected exception'                        
        
  Test-Case-8:
    id: tc_08
    name: make_dir
    desc: Tests for method make_dir
  
    Test-Condition-1: 
      id: tco_01
      name: make_dir
      desc: Make directory
      test: |
        c = ftp.TestFTPClient()
        c._events = []
        path = '/home/hydratk/tmp'
        res = c.make_dir(path)
                         
      validate: |  
        expected = True
        current = res
        this.test_result = (current == expected)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(expected, current) 
        events = ['ftp_before_make_dir']
        this.test_result = (c._events == events)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(events, c._events)                  
        
    Test-Condition-2: 
      id: tco_02
      name: exception
      desc: Make directory exception
      test: |
        res = c.make_dir(path, ex=True)
                         
      validate: |  
        this.test_result = (res == False)
        assert (this.test_result == True), 'Expected exception' 
        
  Test-Case-9:
    id: tc_09
    name: remove_dir
    desc: Tests for method remove_dir
  
    Test-Condition-1: 
      id: tco_01
      name: remove_dir
      desc: Remove directory
      test: |
        c = ftp.TestFTPClient()
        c._events = []
        path = '/home/hydratk/tmp'
        res = c.remove_dir(path)
                         
      validate: |  
        expected = True
        current = res
        this.test_result = (current == expected)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(expected, current) 
        events = ['ftp_before_remove_dir']
        this.test_result = (c._events == events)
        assert (this.test_result == True), 'Expected: {0}, current: {1}'.format(events, c._events)                   
        
    Test-Condition-2: 
      id: tco_02
      name: exception
      desc: Remove directory exception
      test: |
        res = c.remove_dir(path, ex=True)
                         
      validate: |  
        this.test_result = (res == False)
        assert (this.test_result == True), 'Expected exception'                                                                                      